src/source/agc.rs

```diff
@@ -0,0 +1,462 @@
+    /// more accurately while maintaining smoother behavior for gradual changes.
+    #[inline]
+    fn update_peak_level(&mut self, sample_value: f32) {
+        #[cfg(feature = "experimental")]
+        let attack_coeff = if sample_value > self.peak_level {
+            self.attack_coeff
+                .load(Ordering::Relaxed)
+                .min(self.min_attack_coeff) // User-defined attack time limited via release_time
+        } else {
+            self.release_coeff.load(Ordering::Relaxed)
+        };
+
+        #[cfg(not(feature = "experimental"))]
+        let attack_coeff = if sample_value > self.peak_level {
+            self.attack_coeff.min(self.min_attack_coeff) // User-defined attack time limited via release_time
+        } else {
+            self.release_coeff
+        };
+
+        self.peak_level = attack_coeff * self.peak_level + (1.0 - attack_coeff) * sample_value;
+    }
+
+    /// Updates the RMS (Root Mean Square) level using a circular buffer approach.
+    /// This method calculates a moving average of the squared input samples,
+    /// providing a measure of the signal's average power over time.
+    #[inline]
+    fn update_rms(&mut self, sample_value: f32) -> f32 {
+        let squared_sample = sample_value * sample_value;
+        self.rms_window.push(squared_sample);
+        self.rms_window.mean().sqrt()
+    }
+
+    /// Calculate gain adjustments based on peak levels
+    /// This method determines the appropriate gain level to apply to the audio
+    /// signal, considering the peak level.
+    /// The peak level helps prevent sudden spikes in the output signal.
+    #[inline]
+    fn calculate_peak_gain(&self) -> f32 {
+        #[cfg(feature = "experimental")]
+        {
+            if self.peak_level > 0.0 {
+                (self.target_level.load(Ordering::Relaxed) / self.peak_level)
+                    .min(self.absolute_max_gain.load(Ordering::Relaxed))
+            } else {
+                self.absolute_max_gain.load(Ordering::Relaxed)
+            }
+        }
+
+        #[cfg(not(feature = "experimental"))]
+        {
+            if self.peak_level > 0.0 {
+                (self.target_level / self.peak_level).min(self.absolute_max_gain)
+            } else {
+                self.absolute_max_gain
+            }
+        }
+    }
+
+    #[inline]
+    fn process_sample(&mut self, sample: I::Item) -> I::Item {
+        // Convert the sample to its absolute float value for level calculations
+        let sample_value = sample.to_f32().abs();
+
+        // Dynamically adjust peak level using an adaptive attack coefficient
+        self.update_peak_level(sample_value);
+
+        // Calculate the current RMS (Root Mean Square) level using a sliding window approach
+        let rms = self.update_rms(sample_value);
+
+        // Compute the gain adjustment required to reach the target level based on RMS
+        #[cfg(feature = "experimental")]
```

Comment:
lots of #[cfg(not(feature = "experimental"))] makes the algorithm a lot less readable. Have you thought about adding getters? instead of `self.target_level.load(Ordering::Relaxed) / rms` vs `self.target_level / rms` you would have one: `self.target_level() / rms`. Then the getter contains the "noisy" feature stuff.

Reply:
Good idea working on it